botorch/models/contextual_multioutput.py

```diff
@@ -80,8 +81,14 @@ def __init__(
                 for x, y in self.emb_dims
             ]
         )
+        # TODO: Ideally, the transform should be applied automatically within
+        # gpytorch.priors
+        unif = Uniform(0.0, 2.0)
         self.task_covar_module = RBFKernel(
-            ard_num_dims=n_embs, lengthscale_prior=UniformPrior(0.0, 2.0)
+            ard_num_dims=n_embs,
+            lengthscale_prior=UniformPrior(
+                unif.low, unif.high, transform=biject_to(unif.support)
+            ),
```

Comment:
It seems problematic to have a prior here that makes the MLL non-differentiable (non-continuous even) as there is no restriction to the support of the optimizer. I would expect this to result in a bunch of problems (cc @qingfeng10, @bletham).

What we should instead do here is impose an `Interval` constraint that achieves this. Once we do this, then there will be no prior necessary (no prior = uniform). 

So this should be 
```
self.task_covar_module = RBFKernel(
    ard_num_dims=n_embs,
    lengthscale_constraint=Interval(0.0, 2.0),
)
```

where `Interval` is `gpytorch.constraints.Interval`. This would then use the default sigmoid transform. 

The alternative is to specify `transform=None` on the constraint and handle the bounds natively by the scipy optimizer. With L-BFGS-B this works generally better (o/w the optimizer runs off to +/- inf and doesn't really know what to do for lack of gradients). In `fit_gpytorch_torch` we handle this by explicit clamping (or "projecting onto the feasible set" if you want to be all fancy). We extract the bounds for this automatically here: https://github.com/pytorch/botorch/blob/dd17a4c19890e87460dbb50e14456c2f525aeeeb/botorch/optim/numpy_converter.py#L69-L74

```
self.task_covar_module = RBFKernel(
    ard_num_dims=n_embs,
    lengthscale_constraint=Interval(0.0, 2.0, transform=None, inital_value=1.0),
)
```
(the `initial_value` here instructs the optimizer to use 1.0 as the initial condition for MAP estimation).

Reply:
> It seems problematic to have a prior here that makes the MLL non-differentiable (non-continuous even) as there is no restriction to the support of the optimizer.

Thanks for your detailed explanation, @Balandat. My understanding was that the `transform` in `UniformPrior` should be differentiable and will transform from the unconstrained real space (support of the optimizer) to the `Interval(0, 2)`, but you are right that we should directly use the `Interval` constraint here. I'll make the change based on your second suggestion here.